### DISCUSSION TITLE
Provide arguments to custom xcom backend?

### DISCUSSION THREAD

**Discussion by ldacey** [apache/airflow#19868] [Q&A]
I was able to create a custom XCom backend based on an article I read on Medium. Is it possible for me to tweak this further to support supplying _custom_ container names and prefixes per task or operator?

Common pattern is this:
`extract operator (save the raw data on blob storage) >> transform operator (clean and save the data in a pyarrow dataset in a "staging" area, append-only) >> finalize (read from the staging area, drop duplicates and select columns for reporting/visualizations)`

Each task above saves a list of blob paths to xcom. The next operator reads each file in that list and does something with it.

If I could simply return my data from the task function and it would automatically save it in the proper container with the correct folder structure, then that would be great. I want to ensure that all of the data for Client A is in the `client-a` container, and I do not want to lose any "stage" of the transformed data. It seems like I could cut out a lot of the logic in my operators if I could just specify a container name and blob_path to the XCom backend.

```
class WasbXComBackend(BaseXCom):
    PREFIX = "xcom-abfs://"
    CONTAINER_NAME = "airflow-xcom-backend"

    @staticmethod
    def serialize_value(value: Any):
        if isinstance(value, pa.Table):
            hook = AzureBlobHook(wasb_conn_id="azure_blob")
            blob_name = f"data-{guid()}.parquet"
            with pa.BufferOutputStream() as buf:
                pq.write_table(
                    table=value,
                    where=buf,
                    use_dictionary=True,
                    compression="snappy",
                    flavor="spark",
                )
                hook.load_bytes(
                    container_name=WasbXComBackend.CONTAINER_NAME,
                    blob_name=blob_name,
                    buf=buf.getvalue().to_pybytes(),
                )
                value = WasbXComBackend.PREFIX + blob_name
        elif isinstance(value, pd.DataFrame):
            hook = AzureBlobHook(wasb_conn_id="azure_blob")
            blob_name = f"data-{guid()}.csv"
            with io.StringIO() as buf:
                value.to_csv(path_or_buf=buf, index=False)
                hook.load_string(
                    container_name=WasbXComBackend.CONTAINER_NAME,
                    blob_name=blob_name,
                    string_data=buf.getvalue(),
                )
            value = WasbXComBackend.PREFIX + blob_name
        return BaseXCom.serialize_value(value)
```

**Comment by ldacey** [comment]
Hi @turbaszek, I based the snippet above on your article on Medium. Is there a way for me to customize the container_name and prefix at a per task level to ensure the data gets into the right container and "folders"?

**Comment by dstandish** [comment]
@ldacey re your comment on stackoverflow

> Added a more complete example. Since the primary way I save data on Azure is as partitioned parquet datasets, I would hope to be able to control the full path of the dataset, and some arguments like which column the data should be partitioned by etc.

XCom is meant as a way to pass data from one task to another, internal to airflow.  Why do you want to use XCom in this way (i.e. as 
 a mechanism of storing data for external use, e.g. as  partitioned data sitting in s3?  Why don't you specify this in the task config?

**Reply by ldacey** [reply]
> Why do you want to use XCom in this way (i.e. as a mechanism of storing data for external use, e.g. as partitioned data sitting in s3? Why don't you specify this in the task config?

It is the same data whether I save the DataFrame to blob storage using an operator and then push the blob _path_ to XCom (this is what I do today), or return the DataFrame directly (XCom Backend). In both cases, I read input data (say JSON data from the Zendesk REST API), unnest it with pandas json_normalize, and then return the same tabular pandas DataFrame. I feel like I could get rid of a lot of boilerplate code if I could just return a DataFrame and have it save where I need it to be (maybe with XCOM kwargs that I define). 

I suppose I could save the data to a generic container/bucket which hold all XCom backend data and then copy the data to the destination container. The issue I see with that is the data would exist in two locations, or I would need to delete the data in the XCom container which limits my ability to clear historical tasks freely. 

Also, having client data mixed in a single container seems odd and maybe not allowed - I have always used separate containers/buckets with distinct security. There is a difference between saving file paths or max IDs in an XCom database table versus saving actual data which is what happens when you use the XCom backend. I think finely tuned control over this aspect would be useful.

Perhaps I am not using XComs properly though in general. Here is a real example (only edited the client name which is the bucket/container name):


Extract XCom returns the JSON file which was downloaded by the Hook and the cursor (for incremental downloads, basically the latest updated_at timestamp and ticket ID):
```
zendesk_extract  ['raw/zendesk/support/tickets/2021-12-13/20211213T134113.json']
zendesk_cursor  MTYzOTQwMzEyMC4wfHw1NTMyMDM4fA==
```

Transform task reads the JSON XCom above, normalizes it into a pandas DataFrame and ultimately partitions the data in a parquet dataset. The XCom returned from this task is a list of all of the paths which were saved when the data was partitioned (this is an append-only history of every update made to a ticket ID):

```
return_value  ['client-a/staging/zendesk/support/tickets/date_id=20211113/6fe11baa4455462fb6c31257b07424a8.parquet',  'client-a/staging/zendesk/support/tickets/date_id=20211212/d36c27d21b0e41358106259b4a6b6163.parquet', 'client-a/staging/zendesk/support/tickets/date_id=20211213/1e150188eee24cc9ba8b157ab899df4b.parquet']
```

Finalize tasks reads the XCom list from the transform task and then reads all of the data in each unique partition (for example, all files in the 20211212 partition). This data is sorted and duplicates are dropped, the latest version of each row are then saved (overwritten) in a new dataset. The XCom returned is a list of the paths which were saved:
```
return_value  ['client-a/analytics/zendesk/support/tickets/date_id=20211113/20211113.parquet',   'client-a/analytics/zendesk/support/tickets/date_id=20211212/20211212.parquet', 'client-a/analytics/zendesk/support/tickets/date_id=20211213/20211213.parquet']
```
